File: packages/aid-java/src/main/java/org/agentcommunity/aid/Discovery.java
package org.agentcommunity.aid;

import com.fasterxml.jackson.annotation.JsonIgnoreProperties;
import com.fasterxml.jackson.annotation.JsonProperty;
import com.fasterxml.jackson.databind.ObjectMapper;
import java.net.IDN;
import java.net.URI;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.time.Duration;
import java.util.ArrayList;
import java.util.List;
import java.util.Locale;

public final class Discovery {
  private Discovery() {}

  public static final class DiscoveryOptions {
    public String protocol;
    public Duration timeout = Duration.ofSeconds(5);
    public boolean wellKnownFallback = true;
    public Duration wellKnownTimeout = Duration.ofSeconds(2);
    public boolean requireDnssec = false;
  }

  public static final class DiscoveryResult {
    public final AidRecord record;
    public final int ttl;
    public final String queryName;
    public DiscoveryResult(AidRecord record, int ttl, String queryName) {
      this.record = record; this.ttl = ttl; this.queryName = queryName;
    }
  }

  private static String toALabel(String domain) {
    try { return IDN.toASCII(domain); } catch (Exception e) { return domain; }
  }

  // --- DoH Response DTOs for Jackson ---
  @JsonIgnoreProperties(ignoreUnknown = true)
  private static class DoHResponse {
    @JsonProperty("Status")
    public int status;
    @JsonProperty("AD")
    public boolean ad; // Authenticated Data (DNSSEC)
    @JsonProperty("Answer")
    public List<DoHAnswer> answer;
  }

  @JsonIgnoreProperties(ignoreUnknown = true)
  private static class DoHAnswer {
    @JsonProperty("data")
    public String data;
    @JsonProperty("TTL")
    public int ttl;
  }

  static final class RawTxtAnswer {
    final String data;
    final int ttl;
    RawTxtAnswer(String data, int ttl) {
      this.data = data;
      this.ttl = ttl;
    }
  }

  static final class ParsedRecordWithTtl {
    final AidRecord record;
    final int ttl;
    ParsedRecordWithTtl(AidRecord record, int ttl) { this.record = record; this.ttl = ttl; }
  }

  @FunctionalInterface
  interface WellKnownFetcher {
    AidRecord fetch(String domain, Duration timeout);
  }

  static List<String> queryNames(String alabel, String protocol) {
    List<String> names = new ArrayList<>();
    if (protocol != null && !protocol.isEmpty()) {
      names.add(Constants.DNS_SUBDOMAIN + "._" + protocol + "." + alabel);
    }
    names.add(Constants.DNS_SUBDOMAIN + "." + alabel);
    return names;
  }

  static boolean isNoRecordDohStatus(int status) {
    return status == 3;
  }

  private static DoHResponse queryTxtDoH(String fqdn, Duration timeout) {
    String url = "https://cloudflare-dns.com/dns-query?name=" + URI.create("http://x/"+fqdn).getRawPath().substring(3) + "&type=TXT";
    HttpClient http = HttpClient.newBuilder().followRedirects(HttpClient.Redirect.NEVER).connectTimeout(timeout).build();
    HttpRequest req = HttpRequest.newBuilder(URI.create(url)).timeout(timeout).header("Accept", "application/dns-json").GET().build();
    try {
      HttpResponse<String> res = http.send(req, HttpResponse.BodyHandlers.ofString());
      if (res.statusCode() / 100 != 2) throw new AidError("ERR_DNS_LOOKUP_FAILED", "DoH HTTP "+res.statusCode());
      ObjectMapper mapper = new ObjectMapper();
      DoHResponse doh = mapper.readValue(res.body(), DoHResponse.class);
      if (doh.status != 0) {
        if (isNoRecordDohStatus(doh.status)) throw new AidError("ERR_NO_RECORD", "No TXT answers for " + fqdn);
        throw new AidError("ERR_DNS_LOOKUP_FAILED", "DoH status: " + doh.status);
      }
      if (doh.answer == null || doh.answer.isEmpty()) throw new AidError("ERR_NO_RECORD", "No TXT answers for "+fqdn);
      // Clean up quoted string data from DoH response
      for (DoHAnswer ans : doh.answer) {
        if (ans.data != null && ans.data.length() >= 2 && ans.data.startsWith("\"") && ans.data.endsWith("\""))
          ans.data = ans.data.substring(1, ans.data.length()-1);
      }
      return doh;
    } catch (AidError e) { throw e; }
    catch (Exception e) { throw new AidError("ERR_DNS_LOOKUP_FAILED", e.getMessage()); }
  }

  static ParsedRecordWithTtl selectValidRecord(List<RawTxtAnswer> answers, Duration timeout, String queryName, boolean performHandshake) {
    AidError last = null;
    List<ParsedRecordWithTtl> validRecords = new ArrayList<>();
    for (RawTxtAnswer answer : answers) {
      if (!looksLikeAidRecord(answer.data)) continue;
      try {
        AidRecord rec = Parser.parse(answer.data);
        validRecords.add(new ParsedRecordWithTtl(rec, answer.ttl));
      } catch (AidError e) { last = e; }
    }

    if (!validRecords.isEmpty()) {
      String selectedVersion = Constants.SPEC_VERSION_V1;
      for (ParsedRecordWithTtl result : validRecords) {
        if (Constants.SPEC_VERSION_V2.equals(result.record.v)) {
          selectedVersion = Constants.SPEC_VERSION_V2;
          break;
        }
      }
      List<ParsedRecordWithTtl> selectedRecords = new ArrayList<>();
      for (ParsedRecordWithTtl result : validRecords) {
        if (selectedVersion.equals(result.record.v)) {
          selectedRecords.add(result);
        }
      }
      if (selectedRecords.size() > 1) {
        throw new AidError(
            "ERR_INVALID_TXT",
            "Multiple valid " + selectedVersion + " AID records found for " + queryName + "; publish exactly one valid record per queried DNS name");
      }
      ParsedRecordWithTtl selected = selectedRecords.get(0);
      if (performHandshake && selected.record.pka != null) {
        Handshake.performHandshake(selected.record.uri, selected.record.pka, selected.record.kid == null ? "" : selected.record.kid, timeout);
      }
      return selected;
    }

    throw last != null ? last : new AidError("ERR_NO_RECORD", "No valid AID record in TXT answers");
  }

  private static boolean looksLikeAidRecord(String raw) {
    if (raw == null) return false;
    String[] parts = raw.split(";");
    for (String part : parts) {
      String pair = part.trim();
      int idx = pair.indexOf('=');
      if (idx < 0) continue;
      String key = pair.substring(0, idx).trim().toLowerCase(Locale.ROOT);
      if (!"v".equals(key) && !"version".equals(key)) continue;
      String value = pair.substring(idx + 1).trim();
      return value.toLowerCase(Locale.ROOT).matches("aid[0-9]+");
    }
    return false;
  }

  private static ParsedRecordWithTtl parseSingleValid(List<DoHAnswer> answers, Duration timeout, String queryName) {
    List<RawTxtAnswer> rawAnswers = new ArrayList<>();
    for (DoHAnswer answer : answers) {
      rawAnswers.add(new RawTxtAnswer(answer.data, answer.ttl));
    }
    return selectValidRecord(rawAnswers, timeout, queryName, true);
  }

  public static DiscoveryResult discover(String domain, DiscoveryOptions options) {
    if (options == null) options = new DiscoveryOptions();
    String alabel = toALabel(domain);
    List<String> names = queryNames(alabel, options.protocol);

    AidError last = null;
    for (String name : names) {
      try {
        DoHResponse res = queryTxtDoH(name, options.timeout);
        if (options.requireDnssec && !res.ad) {
          throw new AidError("ERR_SECURITY", "DNSSEC validation failed or was not available for " + name);
        }
        ParsedRecordWithTtl p = parseSingleValid(res.answer, options.timeout, name);
        return new DiscoveryResult(p.record, p.ttl, name);
      } catch (AidError e) {
        last = e;
        if (!"ERR_NO_RECORD".equals(e.errorCode)) break;
      }
    }

    return resolveWellKnownFallback(alabel, options, last, (fallbackDomain, timeout) -> WellKnown.fetch(fallbackDomain, timeout, false));
  }

  static DiscoveryResult resolveWellKnownFallback(
      String alabel, DiscoveryOptions options, AidError last, WellKnownFetcher fetcher) {
    if (last != null && isWellKnownFallbackEligible(last)) {
      if (options.requireDnssec) {
        throw new AidError(
            "ERR_SECURITY",
            "DNSSEC is required; .well-known fallback cannot satisfy dnssec=require for " + alabel);
      }
      if (options.wellKnownFallback) {
        AidRecord rec = fetcher.fetch(alabel, options.wellKnownTimeout);
        return new DiscoveryResult(rec, Constants.DNS_TTL_MIN, Constants.DNS_SUBDOMAIN+"."+alabel);
      }
    }
    throw last != null ? last : new AidError("ERR_DNS_LOOKUP_FAILED", "DNS query failed");
  }

  private static boolean isWellKnownFallbackEligible(AidError error) {
    return "ERR_NO_RECORD".equals(error.errorCode) || "ERR_DNS_LOOKUP_FAILED".equals(error.errorCode);
  }
}

File: packages/aid-dotnet/src/Discovery.cs
using System.Globalization;
using System.Net.Http.Json;
using System.Text.Json;

namespace AidDiscovery;

public sealed class DiscoveryOptions
{
    public string? Protocol;
    public TimeSpan Timeout = TimeSpan.FromSeconds(5);
    public bool WellKnownFallback = true;
    public TimeSpan WellKnownTimeout = TimeSpan.FromSeconds(2);
}

public sealed class DiscoveryResult
{
    public required AidRecord Record { get; init; }
    public required int Ttl { get; init; }
    public required string QueryName { get; init; }
    public bool DomainBound { get; init; }
}

public static class Discovery
{
    // Single shared client reused across all DoH lookups. Creating/disposing an HttpClient
    // per call leaves connections in TIME_WAIT and can exhaust ephemeral ports under load.
    // Per-request timeouts are applied via a linked CancellationTokenSource rather than
    // HttpClient.Timeout so the shared instance stays usable across differing timeouts.
    private static readonly HttpClient SharedClient = new(new SocketsHttpHandler
    {
        AllowAutoRedirect = false,
        PooledConnectionLifetime = TimeSpan.FromMinutes(2),
    });

    private static string ToALabel(string domain)
    {
        try { return new IdnMapping().GetAscii(domain); }
        catch { return domain; }
    }

    internal static IReadOnlyList<string> QueryNames(string alabel, string? protocol)
    {
        var names = new List<string>();
        if (!string.IsNullOrEmpty(protocol))
        {
            names.Add($"{Constants.DnsSubdomain}._{protocol}.{alabel}".TrimEnd('.'));
        }
        names.Add($"{Constants.DnsSubdomain}.{alabel}".TrimEnd('.'));
        return names;
    }

    private static async Task<(List<string> txts, int ttl)> QueryTxtDoHAsync(string fqdn, TimeSpan timeout, CancellationToken cancellationToken)
    {
        // Cloudflare DoH JSON endpoint
        var url = $"https://cloudflare-dns.com/dns-query?name={Uri.EscapeDataString(fqdn)}&type=TXT";
        using var timeoutCts = new CancellationTokenSource(timeout);
        using var linked = CancellationTokenSource.CreateLinkedTokenSource(cancellationToken, timeoutCts.Token);
        using var req = new HttpRequestMessage(HttpMethod.Get, url);
        req.Headers.TryAddWithoutValidation("Accept", "application/dns-json");
        using var res = await SharedClient.SendAsync(req, HttpCompletionOption.ResponseHeadersRead, linked.Token).ConfigureAwait(false);
        if (!res.IsSuccessStatusCode)
            throw new AidError(nameof(Constants.ERR_DNS_LOOKUP_FAILED), $"DoH HTTP {(int)res.StatusCode}");
        using var doc = await JsonDocument.ParseAsync(await res.Content.ReadAsStreamAsync(linked.Token).ConfigureAwait(false), cancellationToken: linked.Token).ConfigureAwait(false);
        var root = doc.RootElement;
        if (root.TryGetProperty("Answer", out var answer) && answer.ValueKind == JsonValueKind.Array)
        {
            var txts = new List<string>();
            int ttl = 0;
            foreach (var a in answer.EnumerateArray())
            {
                if (a.TryGetProperty("data", out var dataEl))
                {
                    var data = dataEl.GetString() ?? string.Empty;
                    // Strip surrounding quotes from TXT payload if present
                    if (data.Length >= 2 && data[0] == '"' && data[^1] == '"')
                    {
                        data = data.Substring(1, data.Length - 2);
                    }
                    txts.Add(data);
                    if (ttl == 0 && a.TryGetProperty("TTL", out var ttlEl) && ttlEl.TryGetInt32(out var t)) ttl = t;
                }
            }
            if (txts.Count > 0) return (txts, ttl);
        }
        // Treat as no-record
        throw new AidError(nameof(Constants.ERR_NO_RECORD), $"No TXT answers for {fqdn}");
    }

    private static async Task<(AidRecord, bool)> ParseSingleValid(IEnumerable<string> txts, TimeSpan timeout, string queryName, string? queriedDomain = null, CancellationToken cancellationToken = default)
    {
        AidError? last = null;
        var byVersion = new Dictionary<string, List<AidRecord>>(StringComparer.Ordinal)
        {
            [Constants.SpecVersionV2] = new(),
            [Constants.SpecVersionV1] = new(),
        };
        foreach (var txt in txts)
        {
            try
            {
                var rec = Aid.Parse(txt);
                if (byVersion.TryGetValue(rec.V, out var records))
                {
                    records.Add(rec);
                }
            }
            catch (AidError e) { last = e; }
        }
        foreach (var version in new[] { Constants.SpecVersionV2, Constants.SpecVersionV1 })
        {
            var records = byVersion[version];
            if (records.Count == 0)
            {
                continue;
            }
            if (records.Count > 1)
            {
                throw new AidError(
                    nameof(Constants.ERR_INVALID_TXT),
                    $"Multiple valid {version} AID records found for {queryName}; publish exactly one valid record per queried DNS name"
                );
            }
            var valid = records[0];
            EnforceDepExpiry(valid, queryName);
            bool domainBound = false;
            if (!string.IsNullOrEmpty(valid.Pka))
            {
                domainBound = await Pka.PerformHandshakeAsync(valid.Uri, valid.Pka!, valid.Kid ?? string.Empty, timeout, domain: queriedDomain, cancellationToken: cancellationToken).ConfigureAwait(false);
            }
            return (valid, domainBound);
        }
        throw last ?? new AidError(nameof(Constants.ERR_NO_RECORD), "No valid AID record in TXT answers");
    }

    // Mirrors the TS client: a record whose `dep` (deprecation) date is in the past is rejected
    // as ERR_INVALID_TXT; a future-dated `dep` is a non-fatal warning. Unparseable dates are
    // ignored here (format is already validated during Aid.Parse).
    internal static void EnforceDepExpiry(AidRecord record, string queryName)
    {
        if (string.IsNullOrEmpty(record.Dep))
        {
            return;
        }
        if (!DateTimeOffset.TryParse(record.Dep, CultureInfo.InvariantCulture, DateTimeStyles.AssumeUniversal | DateTimeStyles.AdjustToUniversal, out var depDate))
        {
            return;
        }
        if (depDate < DateTimeOffset.UtcNow)
        {
            throw new AidError(nameof(Constants.ERR_INVALID_TXT), $"Record for {queryName} was deprecated on {record.Dep}");
        }
        Console.Error.WriteLine($"[AID] WARNING: Record for {queryName} is scheduled for deprecation on {record.Dep}");
    }

    public static async Task<DiscoveryResult> DiscoverAsync(string domain, DiscoveryOptions? options = null, CancellationToken cancellationToken = default)
    {
        options ??= new DiscoveryOptions();
        var alabel = ToALabel(domain);

        var names = QueryNames(alabel, options.Protocol);

        AidError? last = null;
        foreach (var name in names)
        {
            try
            {
                var (txts, ttl) = await QueryTxtDoHAsync(name, options.Timeout, cancellationToken).ConfigureAwait(false);
                var (rec, domainBound) = await ParseSingleValid(txts, options.Timeout, name, alabel, cancellationToken).ConfigureAwait(false);
                return new DiscoveryResult { Record = rec, Ttl = ttl, QueryName = name, DomainBound = domainBound };
            }
            catch (AidError e)
            {
                last = e;
                if (e.ErrorCode != nameof(Constants.ERR_NO_RECORD)) break; // stop on non-NO_RECORD
                continue;
            }
        }

        if (options.WellKnownFallback && last is not null && (last.ErrorCode == nameof(Constants.ERR_NO_RECORD) || last.ErrorCode == nameof(Constants.ERR_DNS_LOOKUP_FAILED)))
        {
            var (rec, domainBound) = await WellKnown.FetchAsync(alabel, options.WellKnownTimeout, queriedDomain: alabel, cancellationToken: cancellationToken).ConfigureAwait(false);
            return new DiscoveryResult { Record = rec, Ttl = Constants.DnsTtlMin, QueryName = $"{Constants.DnsSubdomain}.{alabel}", DomainBound = domainBound };
        }
        throw last ?? new AidError(nameof(Constants.ERR_DNS_LOOKUP_FAILED), "DNS query failed");
    }
}
